store/cachekv/store_bench_test.go

```diff
@@ -11,37 +9,87 @@ import (
+// This benchmark measures the time of iterator.Next() when the parent store is blank
+func benchmarkBlankParentIteratorNext(b *testing.B, keysize int) {
+	b.StopTimer()
+
+	mem := dbadapter.Store{DB: dbm.NewMemDB()}
+	kvstore := cachekv.NewStore(mem)
+	// Use a singleton for value, to not waste time computing it
+	value := randSlice(32)
+	// Use simple values for keys, pick a random start,
+	// and take next b.N keys sequentially after.]
+	startKey := randSlice(32)
+
+	// Add 1 to avoid issues when b.N = 1
+	keys := generateSequentialKeys(startKey, b.N+1)
+	for _, k := range keys {
+		kvstore.Set(k, value)
+	}
+
 	b.ReportAllocs()
+	b.StartTimer()
+
+	iter := kvstore.Iterator(keys[0], keys[b.N])
+
+	for _ = iter.Key(); iter.Valid(); iter.Next() {
+		// TODO: Check if we need to ensure this isn't getting deadcode eliminated
+	}
+
+	iter.Close()
+}
+
+// Benchmark setting New keys to a store, where the new keys are in sequence.
+func benchmarkBlankParentAppend(b *testing.B, keysize int) {
+	b.StopTimer()
+
 	mem := dbadapter.Store{DB: dbm.NewMemDB()}
-	cstore := cachekv.NewStore(mem)
-	keys := make([]string, numKVs)
+	kvstore := cachekv.NewStore(mem)
 
-	for i := 0; i < numKVs; i++ {
-		key := make([]byte, 32)
-		value := make([]byte, 32)
+	// Use a singleton for value, to not waste time computing it
+	value := randSlice(32)
+	// Use simple values for keys, pick a random start,
+	// and take next b.N keys sequentially after.
+	startKey := randSlice(32)
 
-		_, _ = rand.Read(key)
-		_, _ = rand.Read(value)
+	keys := generateSequentialKeys(startKey, b.N)
 
-		keys[i] = string(key)
-		cstore.Set(key, value)
+	b.ReportAllocs()
+	b.StartTimer()
+
+	for _, k := range keys {
+		kvstore.Set(k, value)
 	}
+}
 
-	sort.Strings(keys)
+// Benchmark setting New keys to a store, where the new keys are random.
+// the speed of this function does not depend on the values in the parent store
+func benchmarkRandomSet(b *testing.B, keysize int) {
+	b.StopTimer()
 
-	for n := 0; n < b.N; n++ {
-		iter := cstore.Iterator([]byte(keys[0]), []byte(keys[numKVs-1]))
+	mem := dbadapter.Store{DB: dbm.NewMemDB()}
+	kvstore := cachekv.NewStore(mem)
+
+	// Use a singleton for value, to not waste time computing it
+	value := randSlice(32)
+	keys := generateRandomKeys(keysize, b.N)
 
-		for _ = iter.Key(); iter.Valid(); iter.Next() {
-		}
+	b.ReportAllocs()
+	b.StartTimer()
 
-		iter.Close()
+	for _, k := range keys {
+		kvstore.Set(k, value)
 	}
 }
 
-func BenchmarkCacheKVStoreIterator500(b *testing.B)    { benchmarkCacheKVStoreIterator(500, b) }
-func BenchmarkCacheKVStoreIterator1000(b *testing.B)   { benchmarkCacheKVStoreIterator(1000, b) }
-func BenchmarkCacheKVStoreIterator10000(b *testing.B)  { benchmarkCacheKVStoreIterator(10000, b) }
-func BenchmarkCacheKVStoreIterator50000(b *testing.B)  { benchmarkCacheKVStoreIterator(50000, b) }
-func BenchmarkCacheKVStoreIterator100000(b *testing.B) { benchmarkCacheKVStoreIterator(100000, b) }
```

Comment:
Why aren't we testing with various sizes?

Reply:
@ValarDragon can you help get some of these open questions answered. I would like to merge this asap

Reply:
This one was (intended to be) answered by the overall reply at the bottom. 

The benchmarks atm are structured to do `b.N` of each operation, with then no further repetitions. At least for `Next`, this is expected to be a constant time operation / an operation whose size only depends on the parent cache, hence its consistent with the b.N behavior.

This was because the alternative to do things well, would be to make `b.N` separate, independent stores / independent parents, before starting the timer in order to ensure theres no weird loading with the data cache. 

I think we could do size as a separate parameter, but I don't think this is blocking here personally.